packages/@lwc/engine-core/src/framework/api.ts

```diff
@@ -279,7 +279,15 @@ function s(
                         // to the vnode because the current way the diffing algo works, it will replace the original reference
                         // to the host element with a new one. This means the new element will be mounted and immediately unmounted.
                         // Creating a copy of the vnode to preserve a reference to the previous host element.
-                        clonedVNode = { ...vnode, slotAssignment: data.slotAssignment };
+                        if (!isUndefined(vnode.elm)) {
+                            // Clone when the vnode.elm is defined to ensure we don't lost reference to the previous element.
+                            // This is specifically for slot forwarding.
+                            clonedVNode = { ...vnode, slotAssignment: data.slotAssignment };
+                        } else {
+                            // vnode.elm is undefined during initial render.
+                            // We don't need to clone at this point because it doesn't need to be unmounted.
+                            vnode.slotAssignment = data.slotAssignment;
```

Comment:
The main issue was that we were cloning on every opportunity, but when `vnode.elm` is `undefined`, we miss attaching the element to the vnode entirely which causes a cascade of downstream effects.

Cloning should only be done for slot forwarding which is safe because the previous vnodes will be unmounted and discarded immediately as part of vdom diffing.

Reply:
What's the explanation for why we update the slotAssignment here instead? I'm not very familiar with that property.

Reply:
This runs when the light dom slot forwards its slot assignment to the slottee.

For example:

```html
<x-light-forwarded-slot>
    <div slot="forwardedSlot"></div>
</x-light-forwarded-slot>
```

```html
<x-light-named-slot>
    <slot name="forwardedSlot" slot="namedSlot"></slot>
</x-light-named-slot>
<!--- x-light-forwarded-slot --->
```

```html
<template>
    <slot name="namedSlot"></slot>
</template>
<!--- x-light-named-slot --->
```

When the vnode reaches this point in the code, the `div`'s `vnode.slotAssignment` will be `forwardedSlot`, we need to update it to reflect `namedSlot` so that it is slotted into the correct position in `x-light-named-slot`.

We need to update the `slotAssignment` here because we need to get the latest slot assignment so the content is slotted into the correct position.